File: login-server/src/main/java/work/fking/pangya/login/packet/handler/SetNicknamePacketHandler.java
package work.fking.pangya.login.packet.handler;

import io.netty.buffer.ByteBuf;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import work.fking.pangya.login.LoginServer;
import work.fking.pangya.login.Player;
import work.fking.pangya.login.net.ClientLoginPacketHandler;
import work.fking.pangya.login.net.LoginState;
import work.fking.pangya.networking.protocol.ProtocolUtils;

public class SetNicknamePacketHandler implements ClientLoginPacketHandler {

    private static final Logger LOGGER = LogManager.getLogger(SetNicknamePacketHandler.class);

    @Override
    public void handle(LoginServer server, Player player, ByteBuf packet) {
        var nickname = ProtocolUtils.readPString(packet);

        if (player.loginState() != LoginState.SELECTED_NICKNAME) {
            LOGGER.warn("Unexpected login session state, got={}, expected=SELECTED_NICKNAME", player.loginState());
            player.channel().disconnect();
        }
        player.setLoginState(LoginState.SELECTED_NICKNAME);
    }
}

File: login-server/src/main/kotlin/work/fking/pangya/login/packet/handler/SetNicknamePacketHandler.kt
package work.fking.pangya.login.packet.handler

import io.netty.buffer.ByteBuf
import work.fking.pangya.login.LoginServer
import work.fking.pangya.login.Player
import work.fking.pangya.login.net.ClientPacketHandler
import work.fking.pangya.login.net.LoginState
import work.fking.pangya.networking.protocol.readPString

class SetNicknamePacketHandler : ClientPacketHandler {
    override fun handle(server: LoginServer, player: Player, packet: ByteBuf) {
        val nickname = packet.readPString()

        if (player.loginState() != LoginState.SELECTED_NICKNAME) {
            throw IllegalStateException("Unexpected state ${player.loginState()}, expected SELECTED_NICKNAME")
        }
        player.setLoginState(LoginState.SELECTED_NICKNAME)
    }
}
